rag_app/evaluate.py

```diff
@@ -0,0 +1,115 @@
+import typer
+from pathlib import Path
+import json
+from rag_app.models import EvaluationDataItem
+from lancedb import connect
+from typing import List
+from openai import AsyncOpenAI
+from pydantic import BaseModel
+from tqdm.asyncio import tqdm_asyncio as asyncio
+from rag_app.src.chunking import batch_items
+from tenacity import retry, stop_after_attempt, wait_fixed
+from asyncio import run
+from rag_app.models import TextChunk
+from sklearn.metrics import ndcg_score
+import numpy as np
+
+app = typer.Typer()
+
+
+class EmbeddedEvaluationItem(BaseModel):
+    question: str
+    embedding: List[float]
+    chunk_id: str
+
+
+class QueryResult(BaseModel):
+    source: EmbeddedEvaluationItem
+    results: List[TextChunk]
+
+
+@retry(stop=stop_after_attempt(5), wait=wait_fixed(30))
+async def embed_query(queries: List[EvaluationDataItem], client: AsyncOpenAI):
+    query_strings = [query.question for query in queries]
+    embeddings = await client.embeddings.create(
+        input=query_strings, model="text-embedding-3-large", dimensions=256
+    )
+    embeddings = [embedding_object.embedding for embedding_object in embeddings.data]
+
+    return [
+        EmbeddedEvaluationItem(
+            question=query.question, embedding=embedding, chunk_id=query.chunk_id
+        )
+        for query, embedding in zip(queries, embeddings)
+    ]
+
+
+async def embed_test_queries(
+    queries: List[EvaluationDataItem],
+) -> List[EmbeddedEvaluationItem]:
+    client = AsyncOpenAI()
+    batched_queries = batch_items(queries)
+    coros = [embed_query(query_batch, client) for query_batch in batched_queries]
+    return await asyncio.gather(*coros)
+
+
+async def fetch_relevant_results(
+    queries: List[EmbeddedEvaluationItem],
+    db_path: str,
+    table_name: str,
+) -> List[QueryResult]:
+    db = connect(db_path)
+    table = db.open_table(table_name)
+
+    async def query_table(query: EmbeddedEvaluationItem):
+        results = table.search(query.embedding).limit(10).to_pydantic(TextChunk)
+        return QueryResult(results=results, source=query)
+
+    coros = [query_table(query) for query in queries]
+    return await asyncio.gather(*coros)
+
+
+@app.command(help="Evaluate document retrieval")
+def from_json(
```

Comment:
why is the function call from_json?

Reply:
It reads in the queries from a .json file but now that I'm seeing your other comments, I'll change the generation to be a .jsonl file